### PR TITLE
[ANS][TS SDK] add minting name function

### PR DESCRIPTION
**Pull request by angieyth**
### Description
adding a minting name function to the typescript sdk
### Test Plan
TBD

### CODE REVIEW

File: ecosystem/typescript/sdk/src/plugins/ans_client.ts [hunk 1]
```diff
@@ -94,6 +97,33 @@ export class AnsClient {
...
 
+  /**
+   * Mint a new Aptos name
+   *
+   * @param account AptosAccount where collection will be created
+   * @param domainName  Aptos domain name to mint
+   * @param years year duration of the domain name
+   * @returns The hash of the transaction submitted to the API
+   */
+  async mintAptosName(account: AptosAccount, domainName: string, years: AnyNumber = 1): Promise<string> {
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk/src/plugins/ans_client.ts:108]: `AnyNumber` is `bigint | number` - let's use a simple `number` type..

File: ecosystem/typescript/sdk/src/plugins/ans_client.ts [hunk 2]
```diff
@@ -94,6 +97,33 @@ export class AnsClient {
     return this.getAddressByDomainName(domain);
   }
 
+  /**
+   * Mint a new Aptos name
+   *
+   * @param account AptosAccount where collection will be created
+   * @param domainName  Aptos domain name to mint
+   * @param years year duration of the domain name
+   * @returns The hash of the transaction submitted to the API
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk/src/plugins/ans_client.ts:106]:
```suggestion
   * @returns The hash of the pending transaction submitted to the API
```

**Review by 0xmaayan — APPROVED** [review]
looks good. can we add this change to the CHANGELOG also?

File: ecosystem/typescript/sdk/src/plugins/ans_client.ts [hunk 3]
```diff
@@ -94,6 +97,33 @@ export class AnsClient {
...
+   */
+  async mintAptosName(account: AptosAccount, domainName: string, years: AnyNumber = 1): Promise<string> {
+    // check if the name is valid
+    if (domainName.match(nameComponentPattern) === null) {
+      throw new Error(`Name ${domainName} is not valid`);
```
Comment (davidiw) [inline ecosystem/typescript/sdk/src/plugins/ans_client.ts:111]: shouldn't we have something like an ApiError? Libraries shouldn't throw the generic error, that's an error of last resort, afaiu

File: ecosystem/typescript/sdk/src/plugins/ans_client.ts [hunk 4]
```diff
@@ -94,6 +97,33 @@ export class AnsClient {
...
+    if (address !== null) {
+      throw new Error(`Name ${domainName} is not available`);
+    }
+
+    const builder = new TransactionBuilderRemoteABI(this.provider.aptosClient, { sender: account.address() });
+    const rawTxn = await builder.build(`${this.contractAddress}::domains::register_domain`, [], [domainName, years]);
+
+    const bcsTxn = AptosClient.generateBCSTransaction(account, rawTxn);
+    const pendingTransaction = await this.provider.aptosClient.submitSignedBCSTransaction(bcsTxn);
+    return pendingTransaction.hash;
```
Comment (davidiw) [inline ecosystem/typescript/sdk/src/plugins/ans_client.ts:124]: As I was hacking on the Python SDK, it seems to make sense that the hash should probably be a special typed type and not just a string, even if it is just verified to be 32-bytes... or typed but not a string

**Review by banool — APPROVED** [review]
Looks good!

File: ecosystem/typescript/sdk/src/plugins/ans_client.ts [hunk 5]
```diff
@@ -94,6 +96,39 @@ export class AnsClient {
...
+    if (domainName.match(nameComponentPattern) === null) {
+      throw new ApiError(
+          400,
+          `Name ${domainName} is not valid`,
+      );
+    }
+    // check if the name is available
+    const address = await this.getAddressByName(domainName);
+    if (address !== null) {
+      throw new ApiError(
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk/src/plugins/ans_client.ts:118]: do we need it here? `this.getAddressByName` uses `getTableItem` that has `@parseApiError` decorator - I would assumed that this decorator handles api errors

Comment (angieyth) [reply ecosystem/typescript/sdk/src/plugins/ans_client.ts:118]: we throw this error when `this.getAddressByName(domainName)` return an address. it's not an error for `this.getAddressByName(domainName)` but we are not suppose to mint name in that case. Thus we throw error here.

Comment (0xmaayan) [reply ecosystem/typescript/sdk/src/plugins/ans_client.ts:118]: got it. thanks!

**Pull request merged**